src/core/BufferGeometry.js

```diff
@@ -923,6 +923,32 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
 
 			}
 
+			// ShapeBufferGeometry
+
+			if ( data.type === 'ShapeBufferGeometry' ) {
```

Comment:
How about adding this in `ShapeBufferGeometry`'s own `toJSON()`?

```js
ShapeBufferGeometry.prototype.toJSON = function () {

	var data = BufferGeometry.prototype.toJSON.call( this );

	data.shapes = [];

	if ( Array.isArray( shapes ) ) {

		for ( var i = 0, l = shapes.length; i < l; i ++ ) {

			var shape = shapes[ i ];

			data.shapes.push( shape.uuid );

		}

	} else {

		data.shapes.push( shapes.uuid );

	}

	return data;

}
```

I think that should work?

Reply:
Yes. Indeed, it's better to place this code not in `Geometry` or `BufferGeometry` but in `ShapeGeometry`.